Introduce os-dependent doctest tags `# known bug: macos`, `# known bug: linux`

tobiasdiez opened the pull request:
<!-- ^^^^^
Please provide a concise, informative and self-explanatory title.
Don't put issue numbers in there, do this in the PR body below.
For example, instead of "Fixes #1234" use "Introduce new method to calculate 1+1"
-->
<!-- Describe your changes here in detail -->

Extend the `known bug` doctest tag by the possibility to hide these doctest errors only a certain platform. Use this to hide a few known (mostly randomly occurring) bugs. Also convert a few 'random' tags to 'known bug' if random was used in the sense "on a certain architecture, or with a certain dependency, this fails"`.

Considered alternatives:
- A new (perhaps os-dependent) `skip` tag. In most cases, known bug is however better name since these failures point to actual bugs. But some output is merely architecture dependent and in this case a generic `skip` tag would be better.
- A possibility to only skip these failing tests on ci (e.g. by detecting the `CI` env variable, or by explicitly specifying a baseline as in #36678). This has the disadvantage that these known bugs may randomly popup on developers machines without adding any value in this case.
- A new `random failure` or `known failure` tag with essentially the same meaning as `known bug` but limiting it to random occurrence. This idea didn't got much support in #36372, and the randomness can easily be specified by an additional comment.

Setting to blocker to hide #35715 and #36992 which annoyed devs before https://github.com/sagemath/sage/pull/36769#issuecomment-1826973821. Also hides another unreported test failure in `sage_setup/clean.py`. 

<!-- Why is this change required? What problem does it solve? -->
<!-- If this PR resolves an open issue, please link to it here. For example "Fixes #12345". -->
<!-- If your change requires a documentation PR, please link it appropriately. -->

### :memo: Checklist

<!-- Put an `x` in all the boxes that apply. -->
<!-- If your change requires a documentation PR, please link it appropriately -->
<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
<!-- Feel free to remove irrelevant items. -->

- [ ] The title is concise, informative, and self-explanatory.
- [ ] The description explains in detail what this PR is about.
- [ ] I have linked a relevant issue or discussion.
- [ ] I have created tests covering the changes.
- [ ] I have updated the documentation accordingly.

### :hourglass: Dependencies

<!-- List all open PRs that this PR logically depends on
- #12345: short description why this is a dependency
- #34567: ...
-->
- #36923 as the baseline stats are no longer needed with this PR

<!-- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->

tobiasdiez committed:
bugs in geometry/cone occur only on mac

tobiasdiez committed:
Introduce os-dependent known bugs annotation

And use this to hide a few known (mostly randomly occurring) bugs. Also convert a few 'random' annotations to 'known bug' if random was used in the sense "on a certain architecture, or with a certain dependency, this fails"`.

tobiasdiez committed:
Exclude random error in rings/test

mkoeppe commented on src/doc/en/constructions/calculus.rst:
```diff
@@ -387,7 +387,7 @@ are obtained via ``fourier_series_partial_sum``::
     sage: s5 = f.fourier_series_partial_sum(5); s5
     -6/5*sin(10*x)/pi - 2*sin(6*x)/pi - 6*sin(2*x)/pi + 1/2
-    sage: plot(f, (0,pi)) + plot(s5, (x,0,pi), color='red')
+    sage: plot(f, (0,pi)) + plot(s5, (x,0,pi), color='red')  # known bug: macos, random failures
```
We should avoid the word "random" in these comments because it is a keyword in the doctest code with semantics that are orthogonal to what is denoted here.

@kwankyu suggested the word "sporadic" in https://github.com/sagemath/sage/pull/36372#issuecomment-1868124429; I would support that.

tobiasdiez replied:
Done

mkoeppe commented on src/sage/calculus/transforms/dft.py:
```diff
@@ -798,10 +798,10 @@ def idwt(self, other="haar", wavelet_k=2):
             sage: s = IndexedSequence(A,J)
             sage: t = s.dwt()
-            sage: t            # random arch dependent output
+            sage: t            # known bug, arch dependent output
```
I don't think this is a "bug"

tobiasdiez replied:
See PR description.

mkoeppe replied:
Yes, this is not a case that the PR description covers.

mkoeppe commented on src/sage/structure/coerce_actions.pyx:
```diff
@@ -781,7 +781,7 @@ cdef class IntegerMulAction(IntegerAction):
 
             sage: # needs sage.schemes
-            sage: alarm(0.001); 2^(10^7) * P
+            sage: alarm(0.001); 2^(10^7) * P # known bug: macos, random failures, see #35973
```
As I explained in https://github.com/sagemath/sage/pull/36372#issuecomment-1868342056 (see also https://github.com/sagemath/sage/pull/36372#issuecomment-1868386854, https://github.com/sagemath/sage/pull/36372#issuecomment-1742417444), explanations should go into parentheses because this is what our parsing/unparsing code knows to handle.

tobiasdiez replied:
Done

mkoeppe commented on src/sage/doctest/parsing.py:
```diff
@@ -1210,11 +1242,27 @@ def check_and_clear_tag_counts():
-                                     and tag not in available_software)}
-                        if extra:
+                        extra = {
+                            tag
```
this reformatting seems excessive

tobiasdiez committed:
Use sporadic instead of random, put description in parentheses

tobiasdiez committed:
Put one more description into parentheses

mkoeppe commented on src/sage/rings/function_field/drinfeld_modules/morphism.py:
```diff
@@ -540,12 +540,12 @@ def __invert__(self):
-            sage: f = phi.hom(K.random_element())
-            sage: g = ~f
+            sage: f = phi.hom(K.random_element()) # known bug: macos (sporadic ValueError: the input does not define an isogeny)
+            sage: g = ~f # known bug: macos (sporadic ValueError: the input does not define an isogeny)eny
```
```suggestion
            sage: f = phi.hom(K.random_element())  # known bug: macos (sporadic ValueError: the input does not define an isogeny)
            sage: g = ~f  # known bug: macos (sporadic ValueError: the input does not define an isogeny)
```

per codestyle, 2 spaces before the `#` (please fix throughout)

tobiasdiez committed:
Add two spaces before doctest comment

tobiasdiez committed:
fix remaining doctest failures

tobiasdiez committed:
Known bug with sporadic timeouts in FusionRing class is not specific to macos

https://github.com/sagemath/sage/actions/runs/7364453679/job/20044849349

tobiasdiez committed:
Add known bug in fusion rings

tobiasdiez committed:
Add sporadic seg fault in gap

sage -t --random-seed=108132249120197192887853142590823389437 src/sage/libs/gap/element.pyx

tobiasdiez committed:
Gap issues are sadly not yet resolved

tobiasdiez committed:
Add sporadic failure in expect

tobiasdiez committed:
Add known bug of sporadic timeouts in matrix2

tobiasdiez committed:
More known bug tags in matrix2

tobiasdiez committed:
Add known bug in cone.py

https://github.com/sagemath/sage/actions/runs/7473904078/job/20338992903?pr=37039#step:11:7620

tobiasdiez committed:
Known bug in plot happens also on linux

tobiasdiez committed:
Known bugs related to DrinfeldModuleMorphism also happens on linux

https://github.com/sagemath/sage/actions/runs/7506353535/job/20437594453#step:11:7266

tobiasdiez committed:
Remove print command

tobiasdiez committed:
Hide known bug in ell_rational_field.py

tobiasdiez committed:
Hide bug in EllipticCurve_finite_field.py

tobiasdiez committed:
Hide skip test bug in Matrix_gap class

tobiasdiez committed:
Hide known bug in NumberFieldElement_quadratic.pyx

tobiasdiez committed:
Hide bug in linear_tensor_element.pyx

tobiasdiez committed:
Hide more known bugs in gap/element

tobiasdiez committed:
reduce known-test-failures to cover only known source not matched by known bug annotations

tobiasdiez committed:
fix linter

tobiasdiez committed:
Hide known bug in `interfaces/gap`

tobiasdiez committed:
Remove obsolete known bug annotation in `_find_stale_files` function

tobiasdiez committed:
Merge branch 'develop' into random_bugs
